db/similarity.py

```diff
@@ -0,0 +1,239 @@
+                    instruments,
+                    dortmund,
+                    rosamerica,
+                    tzanetakis)
+             VALUES (
+                    :id,
+                    %(values)s)
+        ON CONFLICT (id)
+         DO NOTHING
+    """ % {"values": values_string})
+    connection.execute(query, {'id': id})
+
+
+def count_similarity():
+    # Get total number of submissions in similarity table
+    with db.engine.connect() as connection:
+        query = text("""
+            SELECT COUNT(*)
+              FROM similarity
+        """)
+        result = connection.execute(query)
+        return result.fetchone()[0]
+
+
+def submit_similarity_by_id(id):
+    """Computes similarity metrics for a single recording specified
+    by lowlevel.id, then inserts the metrics as a new row in the
+    similarity table."""
+    try:
+        id = int(id)
+    except ValueError:
+        raise BadDataException('Parameter `id` must be an integer.')
+
+    with db.engine.connect() as connection:
+        # Check that lowlevel submission exists for given id
+        query = text("""
+            SELECT *
+              FROM lowlevel
+             WHERE id = :id
+        """)
+        result = connection.execute(query, {"id": id})
+        if not result.rowcount:
+            raise db.exceptions.NoDataFoundException('No submission for parameter `id`.')
+
+        metrics = []
+        for name in similarity.metrics.BASE_METRICS:
+            metric_cls = similarity.metrics.BASE_METRICS[name]
+            metric = metric_cls(connection)
+            # Check that metric is initialized
+            query = text("""
+                SELECT *
+                  FROM similarity_metrics
+                 WHERE metric = :metric
+            """)
+            result = connection.execute(query, {"metric": metric.name})
+            if not result.rowcount:
+                # Must init metric
+                metric.create()
+
+            try:
+                metric.calculate_stats()
+            except AttributeError:
+                pass
+            metrics.append(metric)
+
+        vectors_info = []
+        for metric, data in get_metrics_data(id, metrics):
+            try:
+                vector = metric.transform(data)
+                isnan = False
+            except ValueError:
+                vector = [None] * metric.length()
+                isnan = True
+            vectors_info.append((metric.name, vector, isnan))
+
+        insert_similarity(connection, id, vectors_info)
+
+
+def submit_similarity_by_mbid(mbid, offset):
+    """Computes similarity metrics for a single recording specified
+    by (mbid, offset) combination, then inserts the metrics as a new
+    row in the similarity table."""
+    id = db.data.get_lowlevel_id(mbid, offset)
+    submit_similarity_by_id(id)
+
+
+# TODO: Write tests for these and for stats, also add docstrings.
+def insert_similarity_meta(metric, hybrid, description, category):
```

Comment:
@alastair for methods such as this and the four below it, which use sql queries for inserting, updating, and altering, how can I verify their success in a unit test?